send-transaction-service/src/send_transaction_service.rs

```diff
@@ -268,26 +328,25 @@ impl SendTransactionService {
 
             match working_bank.get_signature_status_slot(signature) {
                 None => {
-                    // Transaction is unknown to the working bank, it might have been
-                    // dropped or landed in another fork.  Re-send it
-                    info!("Retrying transaction: {}", signature);
-                    result.retried += 1;
-                    transaction_info.retries += 1;
-                    inc_new_counter_info!("send_transaction_service-retry", 1);
-                    let addresses = leader_info.as_ref().map(|leader_info| {
-                        leader_info.get_leader_tpus(config.leader_forward_count)
-                    });
-                    let addresses = addresses
-                        .map(|address_list| {
-                            if address_list.is_empty() {
-                                vec![tpu_address]
-                            } else {
-                                address_list
-                            }
-                        })
-                        .unwrap_or_else(|| vec![tpu_address]);
-                    for address in addresses {
-                        Self::send_transaction(address, &transaction_info.wire_transaction);
+                    let now = Instant::now();
+                    let need_send = transaction_info
+                        .last_sent_time
+                        .map(|last| now.duration_since(last) >= retry_rate)
+                        .unwrap_or(true);
+                    if need_send {
+                        if transaction_info.last_sent_time.is_some() {
+                            // Transaction sent before is unknown to the working bank, it might have been
+                            // dropped or landed in another fork.  Re-send it
+
+                            info!("Retrying transaction: {}", signature);
+                            result.retried += 1;
+                            transaction_info.retries += 1;
+
+                            inc_new_counter_info!("send_transaction_service-retry", 1);
+                        }
+
+                        batched_transactions.insert(*signature);
```

Comment:
this can overfill the requested batch size if there are a lot of pending transactions eligible for retry.  i think we need to nest two loops to honor the requested batch size

Reply:
Done